File: cpp/tests/test_rng.cpp
/**
 * @file test_rng.cpp
 * @brief Тесты для детерминированного RNG (SplitMix64).
 * 
 * Проверяет совместимость с Python реализацией через golden-значения.
 */

#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <algorithm>
#include <cassert>
#include "alife/rng.h"

using namespace alife;

// Флаги для отслеживания результатов тестов
static int tests_passed = 0;
static int tests_failed = 0;

#define TEST(name) void name()
#define RUN_TEST(name) do { \
    std::cout << "Running " << #name << "... "; \
    try { \
        name(); \
        std::cout << "PASSED" << std::endl; \
        tests_passed++; \
    } catch (const std::exception& e) { \
        std::cout << "FAILED: " << e.what() << std::endl; \
        tests_failed++; \
    } \
} while(0)

#define ASSERT_EQ(expected, actual) do { \
    if ((expected) != (actual)) { \
        throw std::runtime_error("Assertion failed: " #expected " == " #actual); \
    } \
} while(0)

#define ASSERT_TRUE(cond) do { \
    if (!(cond)) { \
        throw std::runtime_error("Assertion failed: " #cond); \
    } \
} while(0)

#define ASSERT_FALSE(cond) do { \
    if (cond) { \
        throw std::runtime_error("Assertion failed: !" #cond); \
    } \
} while(0)

#define ASSERT_NEAR(expected, actual, epsilon) do { \
    if (std::abs((expected) - (actual)) > (epsilon)) { \
        throw std::runtime_error("Assertion failed: " #expected " ~= " #actual); \
    } \
} while(0)

// Golden values из Python реализации
// Получены через: python3 -c "from alife.rng import RNG; rng = RNG(seed=42); [print(rng.next_int()) for _ in range(5)]"

// seed=42 golden values
const uint32_t SEED42_NEXT_INT[] = {803958421, 2993090819, 319790930, 239788948, 608707570};
const double SEED42_NEXT_FLOAT[] = {0.1871861566323787, 0.6968832619022578, 0.07445712806656957, 0.055830215103924274, 0.1417257753200829};

// seed=123 golden values
const uint32_t SEED123_NEXT_INT[] = {1658732843, 4033853308, 3547450344, 2425913855, 2014243122};
const double SEED123_NEXT_FLOAT[] = {0.3862038354855031, 0.9392046621069312, 0.8259551469236612, 0.5648270843084902, 0.46897752257063985};

// seed=2026 golden values
const uint32_t SEED2026_NEXT_INT[] = {2442431779, 3979691357, 3454187662, 3584099570, 127467337};
const double SEED2026_NEXT_FLOAT[] = {0.5686729631852359, 0.9265941001940519, 0.8042407366447151, 0.8344882098026574, 0.02967830211855471};

TEST(test_seed42_next_int) {
    RNG rng(42);
    for (int i = 0; i < 5; i++) {
        ASSERT_EQ(SEED42_NEXT_INT[i], rng.next_int());
    }
}

TEST(test_seed42_next_float) {
    RNG rng(42);
    for (int i = 0; i < 5; i++) {
        ASSERT_NEAR(SEED42_NEXT_FLOAT[i], rng.next_float(), 1e-10);
    }
}

TEST(test_seed123_next_int) {
    RNG rng(123);
    for (int i = 0; i < 5; i++) {
        ASSERT_EQ(SEED123_NEXT_INT[i], rng.next_int());
    }
}

TEST(test_seed123_next_float) {
    RNG rng(123);
    for (int i = 0; i < 5; i++) {
        ASSERT_NEAR(SEED123_NEXT_FLOAT[i], rng.next_float(), 1e-10);
    }
}

TEST(test_seed2026_next_int) {
    RNG rng(2026);
    for (int i = 0; i < 5; i++) {
        ASSERT_EQ(SEED2026_NEXT_INT[i], rng.next_int());
    }
}

TEST(test_seed2026_next_float) {
    RNG rng(2026);
    for (int i = 0; i < 5; i++) {
        ASSERT_NEAR(SEED2026_NEXT_FLOAT[i], rng.next_float(), 1e-10);
    }
}

TEST(test_same_seed_same_sequence) {
    RNG rng1(42);
    RNG rng2(42);
    for (int i = 0; i < 100; i++) {
        ASSERT_EQ(rng1.next_int(), rng2.next_int());
    }
}

TEST(test_zero_seed) {
    RNG rng1(0);
    RNG rng2(0);
    for (int i = 0; i < 10; i++) {
        ASSERT_EQ(rng1.next_int(), rng2.next_int());
    }
}

TEST(test_negative_seed) {
    RNG rng1(-42);
    RNG rng2(-42);
    for (int i = 0; i < 10; i++) {
        ASSERT_EQ(rng1.next_int(), rng2.next_int());
    }
}

TEST(test_large_seed) {
    // 2^63 + 12345
    RNG rng1(static_cast<int64_t>(1LL << 63) + 12345);
    RNG rng2(static_cast<int64_t>(1LL << 63) + 12345);
    for (int i = 0; i < 10; i++) {
        ASSERT_EQ(rng1.next_int(), rng2.next_int());
    }
}

TEST(test_next_int_range) {
    RNG rng(42);
    for (int i = 0; i < 1000; i++) {
        uint32_t value = rng.next_int();
        ASSERT_TRUE(value < (1ULL << 32));
    }
}

TEST(test_next_float_range) {
    RNG rng(42);
    for (int i = 0; i < 1000; i++) {
        double value = rng.next_float();
        ASSERT_TRUE(value >= 0.0 && value < 1.0);
    }
}

TEST(test_randint_range) {
    RNG rng(42);
    for (int i = 0; i < 1000; i++) {
        int32_t value = rng.randint(10, 20);
        ASSERT_TRUE(value >= 10 && value <= 20);
    }
}

TEST(test_randint_single_value) {
    RNG rng(42);
    for (int i = 0; i < 10; i++) {
        ASSERT_EQ(42, rng.randint(42, 42));
    }
}

TEST(test_randint_invalid_range) {
    RNG rng(42);
    bool threw = false;
    try {
        rng.randint(10, 5);
    } catch (const std::invalid_argument&) {
        threw = true;
    }
    ASSERT_TRUE(threw);
}

TEST(test_choice_from_vector) {
    RNG rng(42);
    std::vector<int> items = {1, 2, 3, 4, 5};
    for (int i = 0; i < 100; i++) {
        int value = rng.choice(items);
        ASSERT_TRUE(value >= 1 && value <= 5);
    }
}

TEST(test_choice_from_string) {
    RNG rng(42);
    std::string s = "hello";
    for (int i = 0; i < 50; i++) {
        char c = rng.choice(s);
        ASSERT_TRUE(c == 'h' || c == 'e' || c == 'l' || c == 'l' || c == 'o');
    }
}

TEST(test_choice_empty_vector) {
    RNG rng(42);
    std::vector<int> empty;
    bool threw = false;
    try {
        rng.choice(empty);
    } catch (const std::out_of_range&) {
        threw = true;
    }
    ASSERT_TRUE(threw);
}

TEST(test_choice_deterministic) {
    RNG rng1(999);
    RNG rng2(999);
    std::vector<char> items = {'a', 'b', 'c', 'd', 'e'};
    for (int i = 0; i < 50; i++) {
        ASSERT_EQ(rng1.choice(items), rng2.choice(items));
    }
}

TEST(test_gauss_returns_finite) {
    RNG rng(42);
    for (int i = 0; i < 100; i++) {
        double value = rng.gauss();
        ASSERT_TRUE(std::isfinite(value));
    }
}

TEST(test_gauss_mean_approximation) {
    RNG rng(42);
    double mu = 5.0;
    double sigma = 1.0;
    
    std::vector<double> values;
    for (int i = 0; i < 1000; i++) {
        values.push_back(rng.gauss(mu, sigma));
    }
    
    double sum = 0.0;
    for (double v : values) {
        sum += v;
    }
    double mean = sum / values.size();
    
    // Среднее должно быть близко к mu (с допуском для случайности)
    ASSERT_NEAR(mu, mean, 0.5);
}

TEST(test_gauss_invalid_sigma) {
    RNG rng(42);
    bool threw = false;
    try {
        rng.gauss(0.0, 0.0);
    } catch (const std::invalid_argument&) {
        threw = true;
    }
    ASSERT_TRUE(threw);
}

TEST(test_gauss_deterministic) {
    RNG rng1(777);
    RNG rng2(777);
    for (int i = 0; i < 50; i++) {
        ASSERT_EQ(rng1.gauss(), rng2.gauss());
    }
}

TEST(test_shuffle_preserves_elements) {
    RNG rng(42);
    std::vector<int> original = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    std::vector<int> shuffled = original;
    rng.shuffle(shuffled);
    
    std::sort(shuffled.begin(), shuffled.end());
    ASSERT_EQ(original, shuffled);
}

TEST(test_shuffle_deterministic) {
    std::vector<int> original = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
    
    RNG rng1(555);
    RNG rng2(555);
    
    std::vector<int> lst1 = original;
    std::vector<int> lst2 = original;
    
    rng1.shuffle(lst1);
    rng2.shuffle(lst2);
    
    ASSERT_EQ(lst1, lst2);
}

TEST(test_shuffle_empty_list) {
    RNG rng(42);
    std::vector<int> lst;
    rng.shuffle(lst);
    ASSERT_TRUE(lst.empty());
}

TEST(test_shuffle_single_element) {
    RNG rng(42);
    std::vector<int> lst = {42};
    rng.shuffle(lst);
    ASSERT_EQ(lst.size(), 1);
    ASSERT_EQ(lst[0], 42);
}

TEST(test_copy_same_state) {
    RNG rng1(42);
    // Прогоним несколько значений
    for (int i = 0; i < 10; i++) {
        rng1.next_int();
    }
    
    RNG rng2 = rng1.copy();
    
    // Копия должна давать те же значения
    for (int i = 0; i < 100; i++) {
        ASSERT_EQ(rng1.next_int(), rng2.next_int());
    }
}

TEST(test_copy_independent) {
    RNG rng1(42);
    RNG rng2 = rng1.copy();
    
    // Продвигаем только оригинал на несколько шагов
    for (int i = 0; i < 5; i++) {
        rng1.next_int();
    }
    
    // Копия должна оставаться в исходном состоянии
    RNG rng3(42);  // Третий RNG с тем же начальным seed
    
    // Теперь rng2 (копия) должен давать те же значения, что rng3
    for (int i = 0; i < 10; i++) {
        ASSERT_EQ(rng2.next_int(), rng3.next_int());
    }
}

TEST(test_get_set_state) {
    RNG rng1(42);
    for (int i = 0; i < 10; i++) {
        rng1.next_int();
    }
    
    uint64_t state = rng1.get_state();
    
    RNG rng2(0);
    rng2.set_state(static_cast<int64_t>(state));
    
    for (int i = 0; i < 100; i++) {
        ASSERT_EQ(rng1.next_int(), rng2.next_int());
    }
}

TEST(test_uniform_range) {
    RNG rng(42);
    for (int i = 0; i < 1000; i++) {
        double value = rng.uniform(10.0, 20.0);
        ASSERT_TRUE(value >= 10.0 && value <= 20.0);
    }
}

TEST(test_uniform_deterministic) {
    RNG rng1(12345);
    RNG rng2(12345);
    for (int i = 0; i < 50; i++) {
        ASSERT_EQ(rng1.uniform(0.0, 100.0), rng2.uniform(0.0, 100.0));
    }
}

int main() {
    std::cout << "=== RNG Tests ===" << std::endl;
    std::cout << std::endl;
    
    // Golden value tests
    std::cout << "--- Golden Value Tests ---" << std::endl;
    RUN_TEST(test_seed42_next_int);
    RUN_TEST(test_seed42_next_float);
    RUN_TEST(test_seed123_next_int);
    RUN_TEST(test_seed123_next_float);
    RUN_TEST(test_seed2026_next_int);
    RUN_TEST(test_seed2026_next_float);
    std::cout << std::endl;
    
    // Determinism tests
    std::cout << "--- Determinism Tests ---" << std::endl;
    RUN_TEST(test_same_seed_same_sequence);
    RUN_TEST(test_zero_seed);
    RUN_TEST(test_negative_seed);
    RUN_TEST(test_large_seed);
    std::cout << std::endl;
    
    // next_int tests
    std::cout << "--- next_int Tests ---" << std::endl;
    RUN_TEST(test_next_int_range);
    std::cout << std::endl;
    
    // next_float tests
    std::cout << "--- next_float Tests ---" << std::endl;
    RUN_TEST(test_next_float_range);
    std::cout << std::endl;
    
    // randint tests
    std::cout << "--- randint Tests ---" << std::endl;
    RUN_TEST(test_randint_range);
    RUN_TEST(test_randint_single_value);
    RUN_TEST(test_randint_invalid_range);
    std::cout << std::endl;
    
    // choice tests
    std::cout << "--- choice Tests ---" << std::endl;
    RUN_TEST(test_choice_from_vector);
    RUN_TEST(test_choice_from_string);
    RUN_TEST(test_choice_empty_vector);
    RUN_TEST(test_choice_deterministic);
    std::cout << std::endl;
    
    // gauss tests
    std::cout << "--- gauss Tests ---" << std::endl;
    RUN_TEST(test_gauss_returns_finite);
    RUN_TEST(test_gauss_mean_approximation);
    RUN_TEST(test_gauss_invalid_sigma);
    RUN_TEST(test_gauss_deterministic);
    std::cout << std::endl;
    
    // shuffle tests
    std::cout << "--- shuffle Tests ---" << std::endl;
    RUN_TEST(test_shuffle_preserves_elements);
    RUN_TEST(test_shuffle_deterministic);
    RUN_TEST(test_shuffle_empty_list);
    RUN_TEST(test_shuffle_single_element);
    std::cout << std::endl;
    
    // copy tests
    std::cout << "--- copy Tests ---" << std::endl;
    RUN_TEST(test_copy_same_state);
    RUN_TEST(test_copy_independent);
    std::cout << std::endl;
    
    // state tests
    std::cout << "--- get_state/set_state Tests ---" << std::endl;
    RUN_TEST(test_get_set_state);
    std::cout << std::endl;
    
    // uniform tests
    std::cout << "--- uniform Tests ---" << std::endl;
    RUN_TEST(test_uniform_range);
    RUN_TEST(test_uniform_deterministic);
    std::cout << std::endl;
    
    std::cout << "=== Test Summary ===" << std::endl;
    std::cout << "Passed: " << tests_passed << std::endl;
    std::cout << "Failed: " << tests_failed << std::endl;
    
    return tests_failed == 0 ? 0 : 1;
}

File: python/tests/test_rng.py
"""
Тесты для детерминированного RNG (SplitMix64).
"""

import pytest
from alife.rng import RNG


class TestRNGDeterminism:
    """Тесты на детерминированность RNG."""
    
    def test_same_seed_same_sequence(self):
        """Два RNG с одинаковым seed дают одинаковую последовательность."""
        rng1 = RNG(seed=42)
        rng2 = RNG(seed=42)
        
        for _ in range(100):
            assert rng1.next_int() == rng2.next_int()
    
    def test_same_seed_same_floats(self):
        """Два RNG с одинаковым seed дают одинаковые float значения."""
        rng1 = RNG(seed=12345)
        rng2 = RNG(seed=12345)
        
        for _ in range(50):
            assert rng1.next_float() == rng2.next_float()
    
    def test_different_seed_different_sequence(self):
        """RNG с разными seed дают разные последовательности."""
        rng1 = RNG(seed=1)
        rng2 = RNG(seed=2)
        
        # Вероятность совпадения всех 10 значений крайне мала
        matches = sum(1 for _ in range(10) if rng1.next_int() == rng2.next_int())
        assert matches < 10  # Не все должны совпадать
    
    def test_zero_seed(self):
        """Seed=0 должен работать корректно."""
        rng1 = RNG(seed=0)
        rng2 = RNG(seed=0)
        
        for _ in range(10):
            assert rng1.next_int() == rng2.next_int()
    
    def test_negative_seed(self):
        """Отрицательный seed должен работать корректно."""
        rng1 = RNG(seed=-42)
        rng2 = RNG(seed=-42)
        
        for _ in range(10):
            assert rng1.next_int() == rng2.next_int()
    
    def test_large_seed(self):
        """Большой seed должен работать корректно."""
        rng1 = RNG(seed=2**63 + 12345)
        rng2 = RNG(seed=2**63 + 12345)
        
        for _ in range(10):
            assert rng1.next_int() == rng2.next_int()


class TestNextInt:
    """Тесты для метода next_int()."""
    
    def test_next_int_range(self):
        """next_int() возвращает значения в диапазоне [0, 2^32)."""
        rng = RNG(seed=42)
        
        for _ in range(1000):
            value = rng.next_int()
            assert 0 <= value < (1 << 32)
    
    def test_next_int_is_integer(self):
        """next_int() возвращает целое число."""
        rng = RNG(seed=42)
        assert isinstance(rng.next_int(), int)


class TestNextFloat:
    """Тесты для метода next_float()."""
    
    def test_next_float_range(self):
        """next_float() возвращает значения в диапазоне [0.0, 1.0)."""
        rng = RNG(seed=42)
        
        for _ in range(1000):
            value = rng.next_float()
            assert 0.0 <= value < 1.0
    
    def test_next_float_is_float(self):
        """next_float() возвращает float."""
        rng = RNG(seed=42)
        assert isinstance(rng.next_float(), float)


class TestRandint:
    """Тесты для метода randint()."""
    
    def test_randint_range(self):
        """randint(a, b) возвращает значения в диапазоне [a, b]."""
        rng = RNG(seed=42)
        
        for _ in range(1000):
            value = rng.randint(10, 20)
            assert 10 <= value <= 20
    
    def test_randint_single_value(self):
        """randint(a, a) всегда возвращает a."""
        rng = RNG(seed=42)
        
        for _ in range(10):
            assert rng.randint(42, 42) == 42
    
    def test_randint_negative_range(self):
        """randint() работает с отрицательными диапазонами."""
        rng = RNG(seed=42)
        
        for _ in range(100):
            value = rng.randint(-10, -5)
            assert -10 <= value <= -5
    
    def test_randint_mixed_range(self):
        """randint() работает со смешанными диапазонами."""
        rng = RNG(seed=42)
        
        for _ in range(100):
            value = rng.randint(-5, 5)
            assert -5 <= value <= 5
    
    def test_randint_invalid_range(self):
        """randint(a, b) с a > b выбрасывает ValueError."""
        rng = RNG(seed=42)
        
        with pytest.raises(ValueError):
            rng.randint(10, 5)


class TestChoice:
    """Тесты для метода choice()."""
    
    def test_choice_from_list(self):
        """choice() возвращает элемент из списка."""
        rng = RNG(seed=42)
        items = [1, 2, 3, 4, 5]
        
        for _ in range(100):
            value = rng.choice(items)
            assert value in items
    
    def test_choice_from_tuple(self):
        """choice() работает с кортежами."""
        rng = RNG(seed=42)
        items = (10, 20, 30)
        
        for _ in range(50):
            value = rng.choice(items)
            assert value in items
    
    def test_choice_from_string(self):
        """choice() работает со строками."""
        rng = RNG(seed=42)
        s = "hello"
        
        for _ in range(50):
            value = rng.choice(s)
            assert value in s
    
    def test_choice_empty_sequence(self):
        """choice() с пустой последовательностью выбрасывает IndexError."""
        rng = RNG(seed=42)
        
        with pytest.raises(IndexError):
            rng.choice([])
    
    def test_choice_deterministic(self):
        """choice() детерминирован при одинаковом seed."""
        rng1 = RNG(seed=999)
        rng2 = RNG(seed=999)
        items = ['a', 'b', 'c', 'd', 'e']
        
        for _ in range(50):
            assert rng1.choice(items) == rng2.choice(items)


class TestGauss:
    """Тесты для метода gauss()."""
    
    def test_gauss_returns_float(self):
        """gauss() возвращает float."""
        rng = RNG(seed=42)
        assert isinstance(rng.gauss(), float)
    
    def test_gauss_mean_approximation(self):
        """Среднее значение gauss(mu, sigma) приближается к mu."""
        rng = RNG(seed=42)
        mu = 5.0
        sigma = 1.0
        
        values = [rng.gauss(mu, sigma) for _ in range(1000)]
        mean = sum(values) / len(values)
        
        # Среднее должно быть близко к mu (с допуском для случайности)
        assert abs(mean - mu) < 0.5
    
    def test_gauss_invalid_sigma(self):
        """gauss() с sigma <= 0 выбрасывает ValueError."""
        rng = RNG(seed=42)
        
        with pytest.raises(ValueError):
            rng.gauss(sigma=0)
        
        with pytest.raises(ValueError):
            rng.gauss(sigma=-1)
    
    def test_gauss_deterministic(self):
        """gauss() детерминирован при одинаковом seed."""
        rng1 = RNG(seed=777)
        rng2 = RNG(seed=777)
        
        for _ in range(50):
            assert rng1.gauss() == rng2.gauss()


class TestShuffle:
    """Тесты для метода shuffle()."""
    
    def test_shuffle_preserves_elements(self):
        """shuffle() сохраняет все элементы списка."""
        rng = RNG(seed=42)
        original = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
        shuffled = original.copy()
        rng.shuffle(shuffled)
        
        assert sorted(shuffled) == sorted(original)
    
    def test_shuffle_changes_order(self):
        """shuffle() обычно меняет порядок элементов."""
        rng = RNG(seed=42)
        original = list(range(100))
        shuffled = original.copy()
        rng.shuffle(shuffled)
        
        # С вероятностью ~100% порядок изменится для 100 элементов
        assert shuffled != original or len(original) <= 1
    
    def test_shuffle_deterministic(self):
        """shuffle() детерминирован при одинаковом seed."""
        original = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
        
        rng1 = RNG(seed=555)
        rng2 = RNG(seed=555)
        
        lst1 = original.copy()
        lst2 = original.copy()
        
        rng1.shuffle(lst1)
        rng2.shuffle(lst2)
        
        assert lst1 == lst2
    
    def test_shuffle_empty_list(self):
        """shuffle() работает с пустым списком."""
        rng = RNG(seed=42)
        lst = []
        rng.shuffle(lst)
        assert lst == []
    
    def test_shuffle_single_element(self):
        """shuffle() работает со списком из одного элемента."""
        rng = RNG(seed=42)
        lst = [42]
        rng.shuffle(lst)
        assert lst == [42]


class TestCopy:
    """Тесты для метода copy()."""
    
    def test_copy_same_state(self):
        """copy() создаёт RNG с тем же состоянием."""
        rng1 = RNG(seed=42)
        # Прогоним несколько значений
        for _ in range(10):
            rng1.next_int()
        
        rng2 = rng1.copy()
        
        # Копия должна давать те же значения
        for _ in range(100):
            assert rng1.next_int() == rng2.next_int()
    
    def test_copy_independent(self):
        """Копия независима от оригинала после создания."""
        rng1 = RNG(seed=42)
        rng2 = rng1.copy()

        # Продвигаем только оригинал на несколько шагов
        for _ in range(5):
            rng1.next_int()
        
        # Копия должна оставаться в исходном состоянии
        # и давать значения, которые rng1 давал бы, если бы его не продвигали
        rng3 = RNG(seed=42)  # Третий RNG с тем же начальным seed
        
        # Теперь rng2 (копия) должен давать те же значения, что rng3
        for _ in range(10):
            assert rng2.next_int() == rng3.next_int()
        
        # А rng1 уже даёт другие значения (он был продвинут)
        # Проверяем, что последовательности разные
        rng1_vals = [rng1.next_int() for _ in range(10)]
        rng2_vals = [rng2.next_int() for _ in range(10)]
        assert rng1_vals != rng2_vals

class TestGetSetState:
    """Тесты для get_state() и set_state()."""
    
    def test_get_set_state(self):
        """set_state(get_state()) восстанавливает состояние."""
        rng1 = RNG(seed=42)
        for _ in range(10):
            rng1.next_int()
        
        state = rng1.get_state()
        
        rng2 = RNG(seed=0)
        rng2.set_state(state)
        
        for _ in range(100):
            assert rng1.next_int() == rng2.next_int()
